File: tests/org.eclipse.rap.rwt.test/src/org/eclipse/rwt/internal/lifecycle/EncodingUtil_Test.java
/*******************************************************************************
 * Copyright (c) 2009, 2010 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     EclipseSource - initial API and implementation
 ******************************************************************************/
package org.eclipse.rwt.internal.lifecycle;

import java.util.Arrays;

import org.eclipse.rwt.internal.util.EncodingUtil;

import junit.framework.TestCase;

public class EncodingUtil_Test extends TestCase {

  public void testEscapeDoubleQuoted() {
    String stringToEscape = "First line.\nSecond \"middle\" line.\nThird line.";
    String expected = "First line.\nSecond \\\"middle\\\" line.\nThird line.";
    String result = EncodingUtil.escapeDoubleQuoted( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "First line.\nSecond \\middle\\ line.\nThird line.";
    expected = "First line.\nSecond \\\\middle\\\\ line.\nThird line.";
    result = EncodingUtil.escapeDoubleQuoted( stringToEscape );
    assertEquals( expected, result );
  }

  public void testReplaceNewLines() {
    String stringToReplace = "First line.\nSecond line.\nThird line.";
    String expected = "First line.\\nSecond line.\\nThird line.";
    String result = EncodingUtil.replaceNewLines( stringToReplace );
    assertEquals( expected, result );

    stringToReplace = "First line.\rSecond line.\rThird line.";
    expected = "First line.\\nSecond line.\\nThird line.";
    result = EncodingUtil.replaceNewLines( stringToReplace );
    assertEquals( expected, result );

    stringToReplace = "First line.\r\nSecond line.\r\nThird line.";
    expected = "First line.\\nSecond line.\\nThird line.";
    result = EncodingUtil.replaceNewLines( stringToReplace );
    assertEquals( expected, result );

    stringToReplace = "First line.\r\nSecond line.\r\nThird line.";
    expected = "First line. Second line. Third line.";
    result = EncodingUtil.replaceNewLines( stringToReplace, " " );
    assertEquals( expected, result );
  }

  public void testSplitNewlines() {
    String input = "";
    String[] expected = new String[] { "" };
    String[] result = EncodingUtil.splitNewLines( input );
    assertTrue( Arrays.equals( expected, result ) );

    input = "First line.\nSecond line.\nThird line.";
    expected = new String[] { "First line.", "Second line.", "Third line." };
    result = EncodingUtil.splitNewLines( input );
    assertTrue( Arrays.equals( expected, result ) );

    input = "First line.\rSecond line.\rThird line.";
    result = EncodingUtil.splitNewLines( input );
    assertTrue( Arrays.equals( expected, result ) );

    input = "First line.\r\nSecond line.\r\nThird line.";
    result = EncodingUtil.splitNewLines( input );
    assertTrue( Arrays.equals( expected, result ) );

    input = "First line.\r\nSecond line.\r\nThird line.\r\n";
    expected = new String[] { "First line.", "Second line.", "Third line.", "" };
    result = EncodingUtil.splitNewLines( input );
    assertTrue( Arrays.equals( expected, result ) );
  }

  public void testEscapeLeadingTrailingSpaces() {
    String stringToEscape = "  All rights reserved.   ";
    String expected = "&nbsp;&nbsp;All rights reserved.&nbsp;&nbsp;&nbsp;";
    String result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "All rights reserved. ";
    expected = "All rights reserved.&nbsp;";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "  All rights reserved.";
    expected = "&nbsp;&nbsp;All rights reserved.";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "All rights reserved.";
    expected = "All rights reserved.";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = " \n  All rights reserved. \n ";
    expected = "&nbsp;\n  All rights reserved. \n&nbsp;";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "  ";
    expected = "&nbsp;&nbsp;";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "";
    expected = "";
    result = EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape );
    assertEquals( expected, result );
  }

  public void testReplaceWhiteSpaces() {
    String stringToEscape = "test1 test2";
    String expected = "test1 test2";
    String result = EncodingUtil.replaceWhiteSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "test1  test2";
    expected = "test1&nbsp; test2";
    result = EncodingUtil.replaceWhiteSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "test1   test2";
    expected = "test1&nbsp;&nbsp; test2";
    result = EncodingUtil.replaceWhiteSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = " test";
    expected = "&nbsp;test";
    result = EncodingUtil.replaceWhiteSpaces( stringToEscape );
    assertEquals( expected, result );

    stringToEscape = "  test  ";
    expected = "&nbsp; test&nbsp;&nbsp;";
    result = EncodingUtil.replaceWhiteSpaces( stringToEscape );
    assertEquals( expected, result );
  }

  public void testEscapeSpecialCharacters() {
    String stringToEscape = "abc\u2028abc\u2029abc";
    String expected = "abcabcabc";
    String result = EncodingUtil.escapeSpecialCharacters( stringToEscape );
    assertEquals( expected, result );
  }
}

File: tests/org.eclipse.rap.rwt.jstest/js/org/eclipse/rwt/test/tests/EncodingUtilTest.js
/*******************************************************************************
 * Copyright (c) 2011, 2012 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

(function(){

var EncodingUtil = rwt.util.Encoding;

rwt.qx.Class.define( "org.eclipse.rwt.test.tests.EncodingUtilTest", {

  extend : rwt.qx.Object,

  members : {

    //////////////////////////////////////////////////////
    // Tests ported from WidgetLCAUtil_Test#testEscapeTest

    testEscapeTextWithNull : function() {
      try {
        EncodingUtil.escapeText( null, true );
        EncodingUtil.escapeText( null, true );
        fail();
      } catch( e ) {
        // expected
      }
    },

    testEscapeTextNoChanges : function() {
      assertEquals( "Test", EncodingUtil.escapeText( "Test", false ) );
      assertEquals( "Test", EncodingUtil.escapeText( "Test", true ) );
      assertEquals( "", EncodingUtil.escapeText( "", false ) );
      assertEquals( "", EncodingUtil.escapeText( "", true ) );
    },

    testEscapeBrackets : function() {
      assertEquals( "&lt;", EncodingUtil.escapeText( "<", false ) );
      assertEquals( "&gt;", EncodingUtil.escapeText( ">", false ) );
      assertEquals( "&lt;&lt;&lt;", EncodingUtil.escapeText( "<<<", false ) );
      var expected = "&lt;File &gt;";
      assertEquals( expected, EncodingUtil.escapeText( "<File >", false ) );
      assertEquals( expected, EncodingUtil.escapeText( "<File >", true ) );
    },

    testEscapeAmps : function() {
      assertEquals( "&amp;&amp;&amp;File", EncodingUtil.escapeText( "&&&File", false ) );
    },

    testEscapeMnemonics : function() {
      assertEquals( "Open &amp; Close", EncodingUtil.escapeText( "Open && Close", true ) );
      assertEquals( "E&lt;s&gt;ca'pe&quot; &amp; me",
                    EncodingUtil.escapeText( "&E<s>ca'pe\" && me", true ) );
      // Explicitly call it twice to check that _mnemonicFound is reset
      assertEquals( "E&lt;s&gt;ca'pe&quot; &amp; me",
                    EncodingUtil.escapeText( "&E<s>ca'pe\" && me", true ) );
    },

    testEscapeQuotes : function() {
      assertEquals( "&quot;File&quot;", EncodingUtil.escapeText( "\"File\"", false ) );
      assertEquals( "&quot;&quot;File", EncodingUtil.escapeText( "\"\"File", false ) );
      assertEquals( "&quot;File&quot;", EncodingUtil.escapeText( "\"File\"", true ) );
      assertEquals( "&quot;&quot;File", EncodingUtil.escapeText( "\"\"File", true ) );
    },

    testDontEscapeBackslash: function() {
      assertEquals( "Test\\", EncodingUtil.escapeText( "Test\\", false ) );
    },

    testTruncateAtZero : function() {
      assertEquals( String.fromCharCode( 0 ), "\000" );
      assertEquals( "foo ", EncodingUtil.escapeText( "foo \000 bar", false ) );
      assertEquals( "foo", EncodingUtil.escapeText( "foo\000", false ) );
      assertEquals( "", EncodingUtil.escapeText( "\000foo", false ) );
      assertEquals( "&lt;foo", EncodingUtil.escapeText( "<foo\000>", false ) );
      assertEquals( "&lt;foo", EncodingUtil.escapeText( "<foo\000>", true ) );
    },

    /////////////////////////////////////////////////////////
    // Tests ported from EncodingUtilTest#testReplaceNewLines

    testReplaceNewLines : function() {
      var stringToReplace = "First line.\nSecond line.\nThird line.";
      var expected = "First line.\\nSecond line.\\nThird line.";
      assertEquals( expected, EncodingUtil.replaceNewLines( stringToReplace ) );
    },

    testReplaceCarriageReturns : function() {
      var stringToReplace = "First line.\rSecond line.\rThird line.";
      var expected = "First line.\\nSecond line.\\nThird line.";
      assertEquals( expected, EncodingUtil.replaceNewLines( stringToReplace ) );
    },

    testReplaceDOSNewLines : function() {
      var stringToReplace = "First line.\r\nSecond line.\r\nThird line.";
      var expected = "First line.\\nSecond line.\\nThird line.";
      assertEquals( expected, EncodingUtil.replaceNewLines( stringToReplace ) );
    },

    testReplaceNewLinesWithBlank : function() {
      var stringToReplace = "First line.\r\nSecond line.\r\nThird line.";
      var expected = "First line. Second line. Third line.";
      assertEquals( expected, EncodingUtil.replaceNewLines( stringToReplace, " " ) );
    },

    ////////////////////////////////////////////////////////////
    // Tests ported from EncodingUtilTest#testReplaceWhiteSpaces

    testDontReplaceSingleWhiteSpaces : function() {
      var stringToEscape = "test1 test2";
      var expected = "test1 test2";
      assertEquals( expected, EncodingUtil.replaceWhiteSpaces( stringToEscape ) );
    },

    testReplaceDoubleWhiteSpaces : function() {
      var stringToEscape = "test1  test2";
      var expected = "test1&nbsp; test2";
      assertEquals( expected, EncodingUtil.replaceWhiteSpaces( stringToEscape ) );
    },

    testReplaceMultipleWhiteSpaces : function() {
      var stringToEscape = "test1   test2";
      var expected = "test1&nbsp;&nbsp; test2";
      assertEquals( expected, EncodingUtil.replaceWhiteSpaces( stringToEscape ) );
    },

    testReplaceStartingWhiteSpace : function() {
      var stringToEscape = " test";
      var expected = "&nbsp;test";
      assertEquals( expected, EncodingUtil.replaceWhiteSpaces( stringToEscape ) );
    },

    testReplaceStartingAndEndingWhiteSpaces : function() {
      var stringToEscape = "  test  ";
      var expected = "&nbsp; test&nbsp;&nbsp;";
      assertEquals( expected, EncodingUtil.replaceWhiteSpaces( stringToEscape ) );
    },

    /////////////////////////////////////////////////////////////////////
    // Tests ported from EncodingUtilTest#testEscapeLeadingTrailingSpaces

    testEscapeBothLeadingTrailingSpaces : function() {
      var stringToEscape = "    All rights reserved.   ";
      var expected = "&nbsp;&nbsp;&nbsp;&nbsp;All rights reserved.&nbsp;&nbsp;&nbsp;";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeOnlyTrailingSpaces : function() {
      var stringToEscape = "All rights reserved. ";
      var expected = "All rights reserved.&nbsp;";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeOnlyLeadingSpaces : function() {
      var stringToEscape = "  All rights reserved.";
      var expected = "&nbsp;&nbsp;All rights reserved.";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeNoLeadingTrailingSpaces : function() {
      var stringToEscape = "All rights reserved.";
      var expected = "All rights reserved.";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeLeadingTrailingSpacesWithNewLines : function() {
      var stringToEscape = " \n  All rights reserved. \n ";
      var expected = "&nbsp;\n  All rights reserved. \n&nbsp;";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeLeadingTrailingSpacesWithWhitespaceString : function() {
      var stringToEscape = "  ";
      var expected = "&nbsp;&nbsp;";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    testEscapeLeadingTrailingSpacesWithEmptyString : function() {
      var stringToEscape = "";
      var expected = "";
      assertEquals( expected, EncodingUtil.escapeLeadingTrailingSpaces( stringToEscape ) );
    },

    /////////////////////////////////
    // New tests for mnemonic support

    testRenderMnemonicChar : function() {
      var expected = "Pu<span style=\"text-decoration:underline\">s</span>h";
      assertEquals( expected, EncodingUtil.escapeText( "Push", 2 ) );
    }

  }

} );

}());